include/evmc/loader.h

```diff
@@ -0,0 +1,58 @@
+/* EVMC: Ethereum Client-VM Connector API.
+ * Copyright 2018 Pawel Bylica.
+ * Licensed under the MIT License. See the LICENSE file.
+ */
+
+#pragma once
+
+#if __cplusplus
+extern "C" {
+#endif
+
+/** The function pointer type for EVMC create functions. */
+typedef struct evmc_instance* (*evmc_create_fn)(void);
+
+/** Error codes for the EVMC loader. */
+enum evmc_loader_error_code
+{
+    EVMC_ERRC_SUCCESS = 0,
+    EVMC_ERRC_CANNOT_OPEN,
+    EVMC_ERRC_SYMBOL_NOT_FOUND,
+    EVMC_ERRC_INVALID_ARGUMENT,
+};
+
+/**
+ * Dynamically loads the shared object and with an EVM implementation.
+ *
+ * This function tries to open a DLL at the given `filename`. On UNIX-like systems dlopen() function
+ * is used. On Windows LoadLibrary() function is used.
+ *
+ * If the file does not exists or is not a valid shared library the ::EVMC_ERRC_CANNOT_OPEN error
+ * code is signaled and NULL is returned.
+ *
+ * After the DLL is successfully loaded the function tries to find the EVM create function in the
+ * library. The `filename` is used to guess the EVM name and the name of the create function.
+ * The create function name is constructed as following:
+ * - the filename is taken from the path,
+ * - the "lib" prefix and file extension are stripped from the name,
+ * - all "-" are replaced with "_" to construct _full name_,
+ * - the last stem from the _full name_ separated with "_" is taken to construct the _short name_,
```

Comment:
Can you explain this?

Reply:
How about?

> The _full name_ is split by "_" char and the last item is taken to form the _short name_.

Reply:
Can you give an example of a filename having both short and full names?

Reply:
Yeah, example in the comment would be helpful, like `for "libfull-evm-name.so" the full name "evmc_create_full_evm_name" and "evmc_create_name" are checked in the library`.

Reply:
Also. why do we need these complicated rules, can't we just have one hard-coded name for create function in any evmc library?

Reply:
Want to support both static and dynamic linking of the same source.

Reply:
On POSIX we could use symbol alias in the .so library. On Windows it is much harder. In the end I had to just define `evmc_create()` function that executes `evmc_create_interpreter()` and disable the `evmc_create()` when building a static library. This generally gets messy and requires toolchain specific build flags. And I expect that to be even harder for rust or nim. So I decided to provide this naming convention and keep the implementation of it in this single place.